File: test/error-bound/math.jl
# Test the accuracy of twofloat's math functions
# Function inputs/outputs dataset is created by /test/error-bound/math.ts

json = read("$(@__DIR__)/testset/math.json", String)
testset = JSON.parse(json, TestSet; null=NaN)

args_list = testset.argsList
fn_output = testset.fnOutput

coverage = OrderedDict(keys(fn_output) .=> false)
overflow = OrderedDict(keys(fn_output) .=> 0)

println()
@testset verbose = true "Exponentiation ──────────" begin ######################

    @testset "square_1 (EFT)" begin
        _test(Dict(
            "fn" => "square_1",
            "args" => args_list.op1,
            "rel_err_bound" => 0,
            "compute" => x -> big(x)^2,
        ))
    end

    @testset "square_2" begin
        _test(Dict(
            "fn" => "square_2",
            "args" => args_list.op2,
            "rel_err_bound" => 5u^2,
            "compute" => ((xhi, xlo),) -> (big(xhi) + big(xlo))^2
        ))
    end

    @testset "cube_1" begin
        _test(Dict(
            "fn" => "cube_1",
            "args" => args_list.op1,
            "rel_err_bound" => 3u^2/2 + 4u^3,
            "compute" => x -> big(x)^3
        ))
    end

    @testset "cube_2" begin
        _test(Dict(
            "fn" => "cube_2",
            "args" => args_list.op2,
            "rel_err_bound" => 10u^2 + 25u^4,
            "compute" => ((xhi, xlo),) -> (big(xhi) + big(xlo))^3
        ))
    end

    @testset "_linpow_1" begin
        _test(Dict(
            "fn" => "_linpow_1",
            "args" => args_list.op1n,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => (x, n) -> (x, abs(n)),
            "compute" => (x, n) -> big(x)^n
        ))
    end


    @testset "_logpow_1" begin
        _test(Dict(
            "fn" => "_logpow_1",
            "args" => args_list.op1n,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => (x, n) -> (x, abs(n)),
            "compute" => (x, n) -> big(x)^n
        ))
    end

    @testset "_logpowltr" begin
        _test(Dict(
            "fn" => "_logpowltr",
            "args" => args_list.op1n,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => (x, n) -> (x, abs(n)),
            "compute" => (x, n) -> big(x)^n
        ))
    end

    @testset "_linpow_2" begin
        _test(Dict(
            "fn" => "_linpow_2",
            "args" => args_list.op2n,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => ((xhi, xlo), n) -> ((xhi, xlo), abs(n)),
            "compute" => ((xhi, xlo), n) -> (big(xhi) + big(xlo))^n
        ))
    end

    @testset "_logpow_2" begin
        _test(Dict(
            "fn" => "_logpow_2",
            "args" => args_list.op2n,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => ((xhi, xlo), n) -> ((xhi, xlo), abs(n)),
            "compute" => ((xhi, xlo), n) -> (big(xhi) + big(xlo))^n
        ))
    end

    @testset "powint_1" begin
        _test(Dict(
            "fn" => "powint_1",
            "args" => args_list.op1n,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-95,
            "compute" => (x, n) -> big(x)^n,
            "underflow" => (r, (x, n)) -> n < 0 && abs(u^2/r) < ε₀
        ))
    end

    @testset "powint_2" begin
        _test(Dict(
            "fn" => "powint_2",
            "args" => args_list.op2n,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-95,
            "compute" => ((xhi, xlo), n) -> (big(xhi) + big(xlo))^n,
            "underflow" => (r, (x, n)) -> n < 0 && abs(u^2/r) < ε₀
        ))
    end

    @testset "pow_11" begin
        _test(Dict(
            "fn" => "pow_11",
            "args" => args_list.op11,
            "rel_err_bound_uf" => big(2.0)^-78, # 3ε₀
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => (x, p) -> (abs(x), p),
            "compute" => (x, p) -> big(x)^p
        ))
    end

    @testset "pow_12" begin
        _test(Dict(
            "fn" => "pow_12",
            "args" => args_list.op12,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => (x, p) -> (abs(x), p),
            "compute" => (x, (phi, plo)) -> big(x)^(big(phi) + big(plo))
        ))
    end

    @testset "pow_21" begin
        _test(Dict(
            "fn" => "pow_21",
            "args" => args_list.op21,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => ((xhi, xlo), p) -> xhi < 0 ? ((-xhi, -xlo), p) : ((xhi, xlo), p),
            "compute" => ((xhi, xlo), p) -> (big(xhi) + big(xlo))^p
        ))
    end

    @testset "pow_22" begin
        _test(Dict(
            "fn" => "pow_22",
            "args" => args_list.op22,
            "rel_err_bound_uf" => big(2.0)^-78,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => ((xhi, xlo), p) -> xhi < 0 ? ((-xhi, -xlo), p) : ((xhi, xlo), p),
            "compute" => ((xhi, xlo), (phi, plo)) -> (big(xhi) + big(xlo))^(big(phi) + big(plo))
        ))
    end

    @testset "exp_1" begin
        _test(Dict(
            "fn" => "exp_1",
            "args" => args_list.exp1,
            "rel_err_bound" => 10u^2,
            "compute" => x -> exp(big(x))
        ))
    end

    @testset "exp_2" begin
        _test(Dict(
            "fn" => "exp_2",
            "args" => args_list.exp2,
            "rel_err_bound" => 10u^2,
            "compute" => ((xhi, xlo),) -> exp(big(xhi) + big(xlo))
        ))
    end

    @testset "expm1_1" begin
        _test(Dict(
            "fn" => "expm1_1",
            "args" => args_list.exp1,
            "rel_err_bound" => 10u^2,
            "compute" => x -> expm1(big(x))
        ))
    end

    @testset "expm1_2" begin
        _test(Dict(
            "fn" => "expm1_2",
            "args" => args_list.exp2,
            "rel_err_bound" => 10u^2,
            "compute" => ((xhi, xlo),) -> expm1(big(xhi) + big(xlo))
        ))
    end
end

println()
@testset verbose = true "Logarithms ──────────────" begin ######################

    @testset "ln_1" begin
        _test(Dict(
            "fn" => "ln_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => x -> abs(x),
            "compute" => x -> log(big(x))
        ))
    end

    @testset "ln_2" begin
        _test(Dict(
            "fn" => "ln_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => ((xhi, xlo),) -> xhi < 0 ? ((-xhi, -xlo),) : ((xhi, xlo),),
            "compute" => ((xhi, xlo),) -> log(big(xhi) + big(xlo))
        ))
    end

    @testset "log2_1" begin
        _test(Dict(
            "fn" => "log2_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => x -> abs(x),
            "compute" => x -> log2(big(x))
        ))
    end

    @testset "log2_2" begin
        _test(Dict(
            "fn" => "log2_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => ((xhi, xlo),) -> xhi < 0 ? ((-xhi, -xlo),) : ((xhi, xlo),),
            "compute" => ((xhi, xlo),) -> log2(big(xhi) + big(xlo))
        ))
    end

    @testset "log10_1" begin
        _test(Dict(
            "fn" => "log10_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => x -> abs(x),
            "compute" => x -> log10(big(x))
        ))
    end

    @testset "log10_2" begin
        _test(Dict(
            "fn" => "log10_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => ((xhi, xlo),) -> xhi < 0 ? ((-xhi, -xlo),) : ((xhi, xlo),),
            "compute" => ((xhi, xlo),) -> log10(big(xhi) + big(xlo))
        ))
    end
end

println()
@testset verbose = true "Roots ───────────────────" begin ######################

    @testset "sqrt_1" begin
        _test(Dict(
            "fn" => "sqrt_1",
            "args" => args_list.op1,
            "rel_err_bound" => 25u^2/8,
            "process_args" => x -> abs(x),
            "compute" => x -> sqrt(big(x))
        ))
    end

    @testset "sqrt_2" begin
        _test(Dict(
            "fn" => "sqrt_2",
            "args" => args_list.op2,
            "rel_err_bound" => 25u^2/8,
            "process_args" => ((xhi, xlo),) -> xhi < 0 ? ((-xhi, -xlo),) : ((xhi, xlo),),
            "compute" => ((xhi, xlo),) -> sqrt(big(xhi) + big(xlo))
        ))
    end

    @testset "cbrt_1" begin
        _test(Dict(
            "fn" => "cbrt_1",
            "args" => args_list.op1,
            "rel_err_bound" => 25u^2,
            "compute" => x -> cbrt(big(x))
        ))
    end

    @testset "cbrt_2" begin
        _test(Dict(
            "fn" => "cbrt_2",
            "args" => args_list.op2,
            "rel_err_bound" => 25u^2,
            "compute" => ((xhi, xlo),) -> cbrt(big(xhi) + big(xlo))
        ))
    end

    @testset "nthroot_1" begin
        _test(Dict(
            "fn" => "nthroot_1",
            "args" => args_list.op1n,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => (x, n) -> (x < 0 && iseven(n) ? -x : x, abs(n)),
            "compute" => (x, n) -> sign(x) * big(abs(x))^inv(big(n))
        ))
    end

    @testset "nthroot_2" begin
        _test(Dict(
            "fn" => "nthroot_2",
            "args" => args_list.op2n,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => ((xhi,xlo), n) -> (xhi<0 && iseven(n) ? (-xhi,-xlo) : (xhi,xlo), abs(n)),
            "compute" => function((xhi, xlo), n)
                x = big(xhi) + big(xlo)
                sign(x) * big(abs(x))^inv(big(n))
            end
        ))
    end
end

println()
@testset verbose = true "Modular Arithmetic ──────" begin ######################

    @testset "rempi_1" begin
        _test(Dict(
            "fn" => "rempi_1",
            "args" => args_list.op1,
            "rel_err_bound" => 10u^2,
            "compute" => x -> rem(big(x), big(pi), RoundToZero)
        ))
    end

    @testset "rempi_2" begin
        _test(Dict(
            "fn" => "rempi_2",
            "args" => args_list.op2,
            "rel_err_bound" => 25u^2,
            "compute" => ((xhi, xlo),) -> rem(big(xhi) + big(xlo), big(pi), RoundToZero)
        ))
    end
    @testset "rem2pi_1" begin
        _test(Dict(
            "fn" => "rem2pi_1",
            "args" => args_list.op1,
            "rel_err_bound" => 10u^2,
            "compute" => x -> rem2pi(big(x), RoundToZero)
        ))
    end

    @testset "rem2pi_2" begin
        _test(Dict(
            "fn" => "rem2pi_2",
            "args" => args_list.op2,
            "rel_err_bound" => 25u^2,
            "compute" => ((xhi, xlo),) -> rem2pi(big(xhi) + big(xlo), RoundToZero)
        ))
    end
end

##

println()
@testset "Math functions coverage ─" begin
    for (fn, covered) in coverage
        @test (fn, covered) == (fn, true)
    end
end

overflowed = filter(((fn , ov_count),) -> ov_count > 0, overflow)
if !isempty(overflowed)
    println()
    @info ["overflow\n ", (rpad(k, 20, ' ') * "$v\n " for (k, v) in overflowed)...] |> join
end

File: test/error-bound/math.ts
/**
 * @file Pre-test for math functions
 *
 *  - exponentiation
 *  - logarithms
 *  - roots
 *  - hyperbolic fn
 */

import {
  normalize, abs2, add11,
  square_1, square_2, cube_1, cube_2,
  exp_1, exp_2, expm1_1, expm1_2, powint_1, powint_2, pow_11, pow_12, pow_21, pow_22,
  _linpow_1, _logpow_1, _logpowltr, _linpow_2, _logpow_2,
  sqrt_1, sqrt_2, cbrt_1, cbrt_2, nthroot_1, nthroot_2,
  ln_1, ln_2, log2_1, log2_2, log10_1, log10_2,
  rem2pi_1, rem2pi_2, rempi_1, rempi_2,
} from '../../src/index';

import {
  FnSig, UnionToIntersection, randomFn, E_SPLIT_MAX, collectOutputs, initArgsList,
  FnBySigOpt
} from '../utils';

import { exponent } from '@lvlte/ulp';
import fs from 'node:fs';

// Functions to test grouped by signature
const fnBySig = {
  'op1': {square_1, cube_1, sqrt_1, cbrt_1, ln_1, log2_1, log10_1, rempi_1, rem2pi_1},
  'op2': {square_2, cube_2, sqrt_2, cbrt_2, ln_2, log2_2, log10_2, rempi_2, rem2pi_2},
  'op1n': {_linpow_1, _logpow_1, _logpowltr, powint_1, nthroot_1},
  'op2n': {_linpow_2, _logpow_2, powint_2, nthroot_2},
  'exp1': {exp_1, expm1_1},
  'exp2': {exp_2, expm1_2},
  'op11': {pow_11},
  'op12': {pow_12},
  'op21': {pow_21},
  'op22': {pow_22}
} satisfies FnBySigOpt;

type FnBySig = typeof fnBySig;
type TestedFunctions = UnionToIntersection<FnBySig[keyof FnBySig]>;
type FnName = keyof TestedFunctions;
type ArgsListBySig = { [K in keyof FnBySig]: Parameters<FnSig[K]>[] };
type FnOutputList = { [K in FnName]: ReturnType<TestedFunctions[K]>[] };
type FnArgs = { [K in FnName]: Parameters<TestedFunctions[K]> };

// Return the appropriate processArgs callback for the given function. The role
// of the callback is to make the function parameters fit its domain if needed.
function processArgsFn(fnName: FnName): Function {
  switch (fnName) {
    case 'sqrt_1':
    case 'ln_1':
    case 'log2_1':
    case 'log10_1':
    case 'pow_11':
    case 'pow_12':
      return (...args: FnArgs[typeof fnName]) => (args[0] = Math.abs(args[0]), args);

    case 'sqrt_2':
    case 'ln_2':
    case 'log2_2':
    case 'log10_2':
    case 'pow_21':
    case 'pow_22':
      return (...args: FnArgs[typeof fnName]) => (args[0] = abs2(args[0]), args);

    case '_linpow_1':
    case '_logpow_1':
    case '_logpowltr':
    case '_linpow_2':
    case '_logpow_2':
      return (...args: FnArgs[typeof fnName]) => (args[1] = Math.abs(args[1]), args);

    case 'nthroot_1':
      return (...args: FnArgs[typeof fnName]) => {
        const [x, n] = args;
        args[1] = Math.abs(n);
        if (x < 0 && n % 2 === 0) {
          args[0] = -x;
        }
        return args;
      }

    case 'nthroot_2':
      return (...args: FnArgs[typeof fnName]) => {
        const [[xhi, xlo], n] = args;
        args[1] = Math.abs(n);
        if (xhi < 0 && n % 2 === 0) {
          args[0] = [-xhi, -xlo];
        }
        return args;
      }
  }

  return (...args: FnArgs[typeof fnName]) => args;
}

// Pseudo-random number generator
const SEED = Math.sqrt(2);
const random = randomFn(SEED, true);
const rand = randomFn(SEED, false);

// Lists of arguments (grouped by FnSig) to pass to the TestedFunctions
const argsList = initArgsList<ArgsListBySig>(fnBySig);

const emin = -106;
const emax = 53;

for (let exp = emin; exp <= emax; exp++) {
  for (const sign of [1, -1]) {
    for (let r = 0; r < 100; r++) {
      const x = random(exp, sign);
      const y = random(exp - 52, sign);
      const xy = add11(x, y);

      if ([x, y, ...xy].some(v => !Number.isFinite(v))) {
        continue;
      }

      argsList['op1'].push([x]);
      argsList['op2'].push([xy]);
    }

    // power functions
    const exp_max = Math.min(Math.trunc(1022 / Math.abs(exp)), 1022);
    for (let n = 3; n <= exp_max;) {
      for (let r = 0; r < 10; r++) {
        const x = random(exp, sign);
        if (Number.isFinite(x**n) && Math.abs(exponent(x**n)) < E_SPLIT_MAX) {
          const xx = normalize(x, random(exp - 52, -1*sign));
          const exp_sign = r % 2 ? -1 : 1;
          const sn = exp_sign * n;
          const p = exp_sign * (n - rand());
          const pp = normalize(p, random(exponent(p) - 52, 1));
          argsList['op1n'].push([x, sn]);
          argsList['op2n'].push([xx, sn]);
          argsList['op11'].push([x, p]);
          argsList['op12'].push([x, pp]);
          argsList['op21'].push([xx, p]);
          argsList['op22'].push([xx, pp]);
        }
      }
      n = n > 50 ? Math.trunc(n * (1 + random(-2, 1))) : n + 1;
    }
  }
}

// e^x specific input range
// - e^-745.134 < Number.MIN_VALUE
// - e^+709.783 > Number.MAX_VALUE
const e_negx = Math.log2(-Math.log(Number.MIN_VALUE)) - 1;
const e_posx = Math.log2(+Math.log(Number.MAX_VALUE)) - 1;
for (const [sign, emax] of [[1, e_posx], [-1, e_negx]]) {
  const emaxint = Math.floor(emax);
  for (let exp = emin; exp <= emax; exp = exp == emaxint ? emax : exp+1) {
    for (let r = 0; r < 200; r++) {
      const x = random(exp, sign);
      const y = random(exp - 52, sign);
      const xy = add11(x, y);
      argsList['exp1'].push([x]);
      argsList['exp2'].push([xy]);
    }
  }
}

const fnOutput = collectOutputs<FnOutputList>(fnBySig, argsList, processArgsFn);
const testset = { argsList, fnOutput };

// Export as JSON
const testsetJSON = JSON.stringify(testset);
fs.writeFileSync('test/error-bound/testset/math.json', testsetJSON, 'utf8');

console.log('prerun math.ts done');
